crates/bevy_reflect/src/set.rs

```diff
@@ -0,0 +1,557 @@
+    pub fn value_type_path_table(&self) -> &TypePathTable {
+        &self.value_type_path
+    }
+
+    /// The [`TypeId`] of the value.
+    pub fn value_type_id(&self) -> TypeId {
+        self.value_type_id
+    }
+
+    /// Check if the given type matches the value type.
+    pub fn value_is<T: Any>(&self) -> bool {
+        TypeId::of::<T>() == self.value_type_id
+    }
+
+    /// The docstring of this set, if any.
+    #[cfg(feature = "documentation")]
+    pub fn docs(&self) -> Option<&'static str> {
+        self.docs
+    }
+}
+
+/// An ordered set of reflected values.
+#[derive(Default)]
+pub struct DynamicSet {
+    represented_type: Option<&'static TypeInfo>,
+    values: Vec<Box<dyn Reflect>>,
+    indices: HashMap<u64, usize>,
+}
+
+impl DynamicSet {
+    /// Sets the [type] to be represented by this `DynamicSet`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given [type] is not a [`TypeInfo::Set`].
+    ///
+    /// [type]: TypeInfo
+    pub fn set_represented_type(&mut self, represented_type: Option<&'static TypeInfo>) {
+        if let Some(represented_type) = represented_type {
+            assert!(
+                matches!(represented_type, TypeInfo::Set(_)),
+                "expected TypeInfo::Set but received: {:?}",
+                represented_type
+            );
+        }
+
+        self.represented_type = represented_type;
+    }
+
+    /// Inserts a typed value into the set.
+    pub fn insert<V: Reflect>(&mut self, value: V) {
+        self.insert_boxed(Box::new(value));
+    }
+}
+
+impl Set for DynamicSet {
+    fn get(&self, value: &dyn Reflect) -> Option<&dyn Reflect> {
+        self.indices
+            .get(&value.reflect_hash().expect(hash_error!(value)))
+            .map(|index| &**self.values.get(*index).unwrap())
+    }
+
+    fn get_mut(&mut self, value: &dyn Reflect) -> Option<&mut dyn Reflect> {
+        self.indices
+            .get(&value.reflect_hash().expect(hash_error!(value)))
+            .cloned()
+            .map(|index| &mut **self.values.get_mut(index).unwrap())
+    }
+
+    fn get_at(&self, index: usize) -> Option<&dyn Reflect> {
+        self.values.get(index).map(|value| &**value)
+    }
+
+    fn len(&self) -> usize {
+        self.values.len()
+    }
+
+    fn iter(&self) -> SetIter {
+        SetIter::new(self)
+    }
+
+    fn drain(self: Box<Self>) -> Vec<Box<dyn Reflect>> {
+        self.values
+    }
+
+    fn clone_dynamic(&self) -> DynamicSet {
+        DynamicSet {
+            represented_type: self.represented_type,
+            values: self
+                .values
+                .iter()
+                .map(|value| value.clone_value())
+                .collect(),
+            indices: self.indices.clone(),
+        }
+    }
+
+    fn insert_boxed(&mut self, mut value: Box<dyn Reflect>) -> bool {
+        match self
+            .indices
+            .entry(value.reflect_hash().expect(hash_error!(value)))
```

Comment:
This feels pretty susceptible to hash collisions.

Reply:
Does https://github.com/bevyengine/bevy/pull/13014/commits/3c43b67e45888258bf237c7d10e75d0598b2435d check the reflexiveness the way you meant it?

Looking at the first part of the comment now as well

Reply:
Ugghh. I'm really unsure if that is what you meant with the `HashTable` but here we go: https://github.com/bevyengine/bevy/pull/13014/commits/600765bf30325818e4b695d67a76fa8f6a54352f

Can you please give me some feedback on it?

Reply:
> Does [3c43b67](https://github.com/bevyengine/bevy/commit/3c43b67e45888258bf237c7d10e75d0598b2435d) check the reflexiveness the way you meant it?

No, I meant whether the elements being inserted in the set have a reflexive `reflect_partial_eq`, not the set itself.

> Ugghh. I'm really unsure if that is what you meant with the HashTable but here we go: https://github.com/bevyengine/bevy/commit/600765bf30325818e4b695d67a76fa8f6a54352f

Yeah that's what I meant.

Reply:
Ahh I think I got it now! Thanks!